src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceGlyphManager.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Windows.Controls;
+using System.Windows.Media;
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.LanguageServices.Implementation.InheritanceMargin.MarginGlyph;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.InheritanceMargin
+{
+    /// <summary>
+    /// Manager controls all the glyphs of Inheritance Margin in <see cref="InheritanceMarginViewMargin"/>.
+    /// </summary>
+    internal class InheritanceGlyphManager : ForegroundThreadAffinitizedObject, IDisposable
+    {
+        private readonly IWpfTextView _textView;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IStreamingFindUsagesPresenter _streamingFindUsagesPresenter;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IClassificationFormatMap _classificationFormatMap;
+        private readonly IUIThreadOperationExecutor _operationExecutor;
+        private readonly IEditorFormatMap _editorFormatMap;
+        private readonly IAsynchronousOperationListener _listener;
+        private Dictionary<InheritanceMarginGlyph, SnapshotSpan> _glyphToTaggedSpan;
+        private readonly Canvas _glyphsContainer;
+
+        // We want to our glyphs to have the same background color as the glyphs in GlyphMargin.
+        private const string GlyphMarginName = "Indicator Margin";
+
+        // The same width and height as the margin of indicator margin.
+        private const double HeighAndWidthOfTheGlyph = 17;
```

Comment:
"Heigh"

Reply:
also, did you test different windows DPIs, VS zoom levels, and text font sizes here?

Reply:
Yep, I have tested it. But will do it again once I resolve all the comments